File: StructManager.cs
using System.Collections.Generic;

/*
 * Copyright (c) 2018, Joshua Filby <joshua@filby.me>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache
{
	using StructDefinition = OSRSCache.definitions.StructDefinition;
	using StructLoader = OSRSCache.definitions.loaders.StructLoader;
	using StructProvider = OSRSCache.definitions.providers.StructProvider;
	using Archive = OSRSCache.fs.Archive;
	using ArchiveFiles = OSRSCache.fs.ArchiveFiles;
	using FSFile = OSRSCache.fs.FSFile;
	using Index = OSRSCache.fs.Index;
	using Storage = OSRSCache.fs.Storage;
	using Store = OSRSCache.fs.Store;

	public class StructManager : StructProvider
	{
		private readonly Store store;
		private readonly IDictionary<int, StructDefinition> structs = new Dictionary<int, StructDefinition>();

		public StructManager(Store store)
		{
			this.store = store;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void load() throws java.io.IOException
		public virtual void load()
		{
			StructLoader loader = new StructLoader();

			Storage storage = store.Storage;
			Index index = store.getIndex(IndexType.CONFIGS);
			Archive archive = index.getArchive(ConfigType.STRUCT.Id);

			byte[] archiveData = storage.loadArchive(archive);
			ArchiveFiles files = archive.getFiles(archiveData);

			foreach (FSFile f in files.Files)
			{
				StructDefinition def = loader.load(f.FileId, f.Contents);
				structs[f.FileId] = def;
			}
		}

		public virtual IDictionary<int, StructDefinition> Structs
		{
			get
			{
				return new Dictionary<int, StructDefinition>(structs);
			}
		}

		public virtual StructDefinition getStruct(int structId)
		{
			return structs[structId];
		}

		public virtual StructDefinition provide(int structId)
		{
			return getStruct(structId);
		}
	}

}

File: src/main/java/net/runelite/cache/StructManager.cpp
#include "StructManager.h"
#include "IndexType.h"
#include "ConfigType.h"

namespace net::runelite::cache
{
	using IOException = java::io::IOException;
	using Collections = java::util::Collections;
	using HashMap = java::util::HashMap;
	using Map = java::util::Map;
	using StructDefinition = net::runelite::cache::definitions::StructDefinition;
	using StructLoader = net::runelite::cache::definitions::loaders::StructLoader;
	using StructProvider = net::runelite::cache::definitions::providers::StructProvider;
	using Archive = net::runelite::cache::fs::Archive;
	using ArchiveFiles = net::runelite::cache::fs::ArchiveFiles;
	using FSFile = net::runelite::cache::fs::FSFile;
	using Index = net::runelite::cache::fs::Index;
	using Storage = net::runelite::cache::fs::Storage;
	using Store = net::runelite::cache::fs::Store;

	StructManager::StructManager(std::shared_ptr<Store> store) : store(store)
	{
	}

	void StructManager::load()
	{
		std::shared_ptr<StructLoader> loader = std::make_shared<StructLoader>();

		std::shared_ptr<Storage> storage = store->getStorage();
		std::shared_ptr<Index> index = store->getIndex(IndexType::CONFIGS);
		std::shared_ptr<Archive> archive = index->getArchive(ConfigType::STRUCT.getId());

		std::vector<signed char> archiveData = storage->loadArchive(archive);
		std::shared_ptr<ArchiveFiles> files = archive->getFiles(archiveData);

		for (auto f : files->getFiles())
		{
			std::shared_ptr<StructDefinition> def = loader->load(f->getFileId(), f->getContents());
			structs.emplace(f->getFileId(), def);
		}
	}

	std::unordered_map<int, std::shared_ptr<StructDefinition>> StructManager::getStructs()
	{
		return Collections::unmodifiableMap(structs);
	}

	std::shared_ptr<StructDefinition> StructManager::getStruct(int structId)
	{
		return structs[structId];
	}

	std::shared_ptr<StructDefinition> StructManager::provide(int structId)
	{
		return getStruct(structId);
	}
}
